helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java

```diff
@@ -358,9 +361,10 @@ protected Map<String, ResourceAssignment> computeBestPossibleAssignment(
       throws HelixRebalanceException {
     // Perform global rebalance for a new baseline assignment
     globalRebalance(clusterData, resourceMap, currentStateOutput, algorithm);
-    // Perform partial rebalance for a new best possible assignment
+    // Perform emergency rebalance for a new best possible assignment
```

Comment:
Q: looks like this is for case(1)?

Reply:
Both cases start with emergency rebalance - it's an alternative to the original design, which I think simplifies things.